tests/NuGet.Services.AzureSearch.Tests/IndexActionBuilderFacts.cs

```diff
@@ -0,0 +1,455 @@
+            [InlineData("nuget.versioning", "1.0.0", "nuget_versioning_1_0_0-bnVnZXQudmVyc2lvbmluZy8xLjAuMA2")]
+            [InlineData("NUGET.VERSIONING", "1.0.0", "nuget_versioning_1_0_0-bnVnZXQudmVyc2lvbmluZy8xLjAuMA2")]
+            [InlineData("_", "1.0.0", "1_0_0-Xy8xLjAuMA2")]
+            [InlineData("foo-bar", "1.0.0", "foo-bar_1_0_0-Zm9vLWJhci8xLjAuMA2")]
+            [InlineData("İzmir", "1.0.0", "zmir_1_0_0-xLB6bWlyLzEuMC4w0")]
+            [InlineData("İİzmir", "1.0.0", "zmir_1_0_0-xLDEsHptaXIvMS4wLjA1")]
+            [InlineData("zİİmir", "1.0.0", "z__mir_1_0_0-esSwxLBtaXIvMS4wLjA1")]
+            [InlineData("zmirİ", "1.0.0", "zmir__1_0_0-em1pcsSwLzEuMC4w0")]
+            [InlineData("zmirİİ", "1.0.0", "zmir___1_0_0-em1pcsSwxLAvMS4wLjA1")]
+            [InlineData("惡", "1.0.0", "1_0_0-5oOhLzEuMC4w0")]
+            [InlineData("jQuery", "1.0.0-alpha", "jquery_1_0_0-alpha-anF1ZXJ5LzEuMC4wLWFscGhh0")]
+            [InlineData("jQuery", "1.0.0-Alpha", "jquery_1_0_0-alpha-anF1ZXJ5LzEuMC4wLWFscGhh0")]
+            [InlineData("jQuery", "1.0.0-ALPHA", "jquery_1_0_0-alpha-anF1ZXJ5LzEuMC4wLWFscGhh0")]
+            [InlineData("jQuery", "1.0.0.0-ALPHA", "jquery_1_0_0-alpha-anF1ZXJ5LzEuMC4wLWFscGhh0")]
+            [InlineData("jQuery", "01.0.0-ALPHA", "jquery_1_0_0-alpha-anF1ZXJ5LzEuMC4wLWFscGhh0")]
+            [InlineData("jQuery", "1.0.0-ALPHA+githash", "jquery_1_0_0-alpha-anF1ZXJ5LzEuMC4wLWFscGhh0")]
+            [InlineData("jQuery", "1.0.0-ALPHA.1+githash", "jquery_1_0_0-alpha_1-anF1ZXJ5LzEuMC4wLWFscGhhLjE1")]
+            [InlineData("jQuery", "1.0.0-ALPHA.1", "jquery_1_0_0-alpha_1-anF1ZXJ5LzEuMC4wLWFscGhhLjE1")]
+            public void EncodesHijackDocumentKey(string id, string version, string expected)
+            {
+                var input = new NewPackageRegistration(
+                    id,
+                    1001,
+                    new string[0],
+                    new[] { new TestPackage(version) });
+
+                var actions = _target.AddNewPackageRegistration(input);
+
+                Assert.Equal(1, actions.Hijack.Count);
+                Assert.Equal(expected, actions.Hijack[0].Document.Key);
+            }
+        }
+
+        private static async Task<string> SerializeToJsonAsync<T>(T obj) where T : class
+        {
+            using (var testHandler = new TestHttpClientHandler())
+            using (var serviceClient = new SearchServiceClient(
+                "unit-test-service",
+                new SearchCredentials("unit-test-api-key"),
+                testHandler))
+            {
+                var indexClient = serviceClient.Indexes.GetClient("unit-test-index");
+                await indexClient.Documents.IndexAsync(IndexBatch.Upload(new[] { obj }));
+                return testHandler.LastRequestBody;
+            }
+        }
+
+        private class TestHttpClientHandler : HttpClientHandler
```

Comment:
The encoding APIs used by Azure Search to take a document and serialize it as JSON before sending it to the REST APIs is private so I made this test HTTP client handler to capture the POSTed JSON.